tortoise/verifying_tortoise.go

```diff
@@ -0,0 +1,602 @@
+			continue
+		}
+		for _, id := range ids {
+			idx, ok := t.BlocksToBlocksIndex[id]
+			if !ok {
+				continue
+			}
+			delete(t.BlocksToBlocksIndex, id)
+			t.BlocksToBlocks = removeOpinion(t.BlocksToBlocks, idx)
+			// FIX indexes
+			for i, op := range t.BlocksToBlocks {
+				t.BlocksToBlocksIndex[op.BlockID] = i
+			}
+			goodidx, ok := t.GoodBlocksIndex[id]
+			if !ok {
+				continue
+			}
+			delete(t.GoodBlocksIndex, id)
+			t.GoodBlocksArr = removeBlockIndex(t.GoodBlocksArr, goodidx)
+			for i, gb := range t.GoodBlocksArr {
+				t.GoodBlocksIndex[gb] = i
+			}
+			t.logger.Debug("evict block %v from maps ", id)
+		}
+	}
+	t.Evict = window
+}
+
+func removeBlockIndex(b []types.BlockID, i int) []types.BlockID {
+	return append(b[:i], b[i+1:]...)
+}
+
+func removeOpinion(o []opinion, i int) []opinion {
+	return append(o[:i], o[i+1:]...)
+}
+
+//func (t *turtle) oldInputVector(l types.LayerID, b types.BlockID) vec {
+//	v, err := t.bdp.GetLayerInputVector(l)
+//	if err != nil {
+//		return abstain
+//	}
+//
+//
+//
+//	calc := t.inputVectorForLayer()
+//
+//	return vi.Vec()
+//}
+
+//func (t *turtle) singleInputVector(l types.LayerID, b types.BlockID) vec {
+//	if l == 0 {
+//		return support
+//	}
+//
+//	// TODO: Pull these from db/sync if we are syncing
+//	res, err := t.hrp.GetResult(l)
+//	if err != nil {
+//		return abstain
+//	}
+//
+//	m := make(map[types.BlockID]vec, len(res))
+//	wasIncluded := false
+//
+//	for _, bl := range res {
+//		m[bl] = support
+//		if bl == b {
+//			wasIncluded = true
+//		}
+//	}
+//
+//	if !wasIncluded {
+//		return against
+//	}
+//
+//	return support
+//}
+
+func (t *turtle) singleInputVectorFromDB(lyrid types.LayerID, blockid types.BlockID) (vec, error) {
+	if lyrid == 0 {
+		return support, nil
+	}
+
+	input, err := t.bdp.GetLayerInputVector(lyrid)
+	if err != nil {
+		return abstain, err
+	}
+
+	m := make(map[types.BlockID]vec, len(input))
+	wasIncluded := false
+
+	for _, bl := range input {
+		m[bl] = support
+		if bl == blockid {
+			wasIncluded = true
+		}
+	}
+
+	if !wasIncluded {
+		return against, nil
+	}
+
+	return support, nil
+}
+
+func (t *turtle) inputVectorForLayer(lyrBlocks []types.BlockID, inputvector *[]types.BlockID) map[types.BlockID]vec {
+	lyrResult := make(map[types.BlockID]vec, len(lyrBlocks))
+	//XXX : input vector must be pointer so we can differentiate
+	// no support votes to no results at all.
+	if inputvector == nil {
+		// hare didn't finish hence we don't have opinion
+		// TODO: get hare opinion when hare finishes
+		for _, b := range lyrBlocks {
+			lyrResult[b] = abstain
+		}
+		return lyrResult
+	}
+
+	input := *inputvector
+	for _, b := range input {
+		lyrResult[b] = support
+	}
+
+	for _, b := range lyrBlocks {
+		if _, ok := lyrResult[b]; !ok {
+			lyrResult[b] = against
+		}
+	}
+	return lyrResult
+}
+
+func (t *turtle) BaseBlock(getres func(id types.LayerID) ([]types.BlockID, error)) (types.BlockID, [][]types.BlockID, error) {
```

Comment:
should this be exported?

Reply:
yes it is used by the block builder to point on a baseblock.